File: frontend/src/api/push.ts
import { apiRequest } from './http';

/** Web Push subscription payload sent to the backend. */
export type PushSubscriptionPayload = {
  endpoint: string;
  p256dh: string;
  auth: string;
};

/** Fetch the server VAPID public key for pushManager.subscribe. */
export async function fetchVAPIDPublicKey(): Promise<string> {
  const res = await apiRequest<{ public_key: string }>('/v1/push/vapid', { method: 'GET' });
  return res.public_key ?? '';
}

/** Register the current browser subscription with the backend. */
export async function registerPushSubscription(sub: PushSubscriptionPayload): Promise<void> {
  await apiRequest('/v1/push/subscriptions', { method: 'POST', body: sub });
}

/** Remove a subscription from the backend (idempotent). */
export async function unregisterPushSubscription(endpoint: string): Promise<void> {
  await apiRequest('/v1/push/subscriptions', { method: 'DELETE', body: { endpoint } });
}

File: backend/internal/handler/push.go
package handler

import (
	"encoding/json"
	"net/http"

	"easy-im/backend/internal/apperr"
	"easy-im/backend/internal/service"
)

// PushHandler serves /v1/push*.
type PushHandler struct {
	Push           *service.PushService
	VAPIDPublicKey string
}

type pushSubscriptionBody struct {
	Endpoint string `json:"endpoint"`
	P256DH   string `json:"p256dh"`
	Auth     string `json:"auth"`
}

// VAPID returns the public key for the browser to subscribe with.
func (h *PushHandler) VAPID(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{"public_key": h.VAPIDPublicKey})
}

// Register stores the caller's Web Push subscription (cookie-authenticated).
func (h *PushHandler) Register(w http.ResponseWriter, r *http.Request) {
	if h.Push == nil {
		WriteError(w, r, apperr.Unavailable("push not configured"))
		return
	}
	var body pushSubscriptionBody
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		WriteError(w, r, apperr.Invalid("invalid request body"))
		return
	}
	if err := h.Push.Register(r.Context(), UserIDFromContext(r.Context()), body.Endpoint, body.P256DH, body.Auth); err != nil {
		WriteError(w, r, err)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// Unregister removes the caller's Web Push subscription (idempotent).
func (h *PushHandler) Unregister(w http.ResponseWriter, r *http.Request) {
	if h.Push == nil {
		WriteError(w, r, apperr.Unavailable("push not configured"))
		return
	}
	var body pushSubscriptionBody
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		WriteError(w, r, apperr.Invalid("invalid request body"))
		return
	}
	if err := h.Push.Unregister(r.Context(), UserIDFromContext(r.Context()), body.Endpoint); err != nil {
		WriteError(w, r, err)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}
